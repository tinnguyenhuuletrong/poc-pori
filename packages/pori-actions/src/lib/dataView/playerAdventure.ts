import type Realm from 'realm';
import {
  IdleGameSCEventDataModel,
  DataViewRepo,
  DataViewModel,
  PoriRepo,
} from '@pori-and-friends/pori-repositories';
import {
  AdventureFinishedData,
  AdventureFortifiedData,
  AdventureInfo,
  AdventureInfoEx,
  AdventureStartedData,
  AdventureStatsGroupByDay,
  AdventureSupported1Data,
  AdventureSupported2Data,
  calculateMineTurnTime,
  EIdleGameSCEventType,
} from '@pori-and-friends/pori-metadata';
import moment from 'moment';

const TOKEN_DECIMAL = 18;
const TOKEN_CONVERT_CONSTANT = 10 ** TOKEN_DECIMAL;

const HEAD_VERSION = '1';
export interface IComputePlayerAdventureProps {
  realm: Realm;
  playerAddress: string;
  realmEventStore: Realm.Results<IdleGameSCEventDataModel & Realm.Object>;
}

export interface ActiveAdventureComputed {
  playerAddress: string;
  note: {
    lastMine?: number;
    lastMineUnlockAt?: Date;
  };
  activeAdventures: Record<number, AdventureInfo>;
  finishedAdventures: Record<number, AdventureStatsGroupByDay>;
}

export async function computePlayerAdventure(
  options: IComputePlayerAdventureProps
) {
  const { realm, playerAddress } = options;
  const viewId = computeViewId(options);
  const lastViewIns = await getLastView(options);
  const { cursor } = lastViewIns;
  const viewData = JSON.parse(lastViewIns.data) as ActiveAdventureComputed;

  const allEvents = options.realmEventStore.filtered(
    `
      _id > oid(${cursor}) && (
        type="AdventureStarted" ||
        type="AdventureSupported1" ||
        type="AdventureFortified" ||
        type="AdventureSupported2" ||
        type="AdventureFinished"
      )
    `
  );

  if (allEvents.length <= 0) return viewData;
  console.log('need to process', { playerAddress, count: allEvents.length });
  for (const it of allEvents) {
    switch (it.type) {
      case EIdleGameSCEventType.AdventureStarted:
        {
          const evData = it.data as AdventureStartedData;
          const isFarmer = evData.farmer === playerAddress;

          viewData.activeAdventures[evData.mineId] = {
            mineId: evData.mineId,
            state: 'AdventureStarted',

            isFarmer,
            farmerAddress: evData.farmer,
            startTime: new Date(evData.startTime * 1000),
            farmerEndTime: new Date(evData.blockedTime * 1000),

            farmerPories: [...evData.porians],
            farmerRewardLevel: [...evData.rewardLevels],
            farmerSlots: [...evData.indexes],
          };
        }
        break;

      case EIdleGameSCEventType.AdventureSupported1:
        {
          const evData = it.data as AdventureSupported1Data;
          const isSupporter = evData.helper === playerAddress;
          const mineInfo = viewData.activeAdventures[evData.mineId];

          if (!mineInfo) {
            console.log('missing', evData.mineId);
            break;
          }

          // not interesting
          if (!(isSupporter || mineInfo.isFarmer)) {
            delete viewData.activeAdventures[evData.mineId];
            break;
          }

          viewData.activeAdventures[evData.mineId] = {
            ...mineInfo,
            state: 'AdventureSupported1',
            isSupporter,
            supporterEndTime: new Date(evData.blockedTime * 1000),
            supporterAddress: evData.helper,

            supporterPories: evData.porians,
            supporterRewardLevel: evData.rewardLevels,
            supporterSlots: evData.indexes,
          };
        }
        break;

      case EIdleGameSCEventType.AdventureFortified:
        {
          const evData = it.data as AdventureFortifiedData;
          const mineInfo = viewData.activeAdventures[evData.mineId];

          // not interesting
          if (!mineInfo) break;
          if (!(mineInfo.isSupporter || mineInfo.isFarmer)) {
            delete viewData.activeAdventures[evData.mineId];
            break;
          }

          viewData.activeAdventures[evData.mineId] = {
            ...mineInfo,
            state: 'AdventureFortified',

            farmerEndTime: new Date(evData.blockedTime * 1000),

            // add more pori
            farmerPories: [...(mineInfo.farmerPories ?? []), evData.porian],
            farmerRewardLevel: [
              ...(mineInfo.farmerRewardLevel ?? []),
              evData.rewardLevel,
            ],
            farmerSlots: [...(mineInfo.farmerSlots ?? []), evData.index],
          };
        }
        break;

      case EIdleGameSCEventType.AdventureSupported2:
        {
          const evData = it.data as AdventureSupported2Data;
          const mineInfo = viewData.activeAdventures[evData.mineId];

          // not interesting
          if (!mineInfo) break;
          if (!(mineInfo.isSupporter || mineInfo.isFarmer)) {
            delete viewData.activeAdventures[evData.mineId];
            break;
          }

          viewData.activeAdventures[evData.mineId] = {
            ...mineInfo,
            state: 'AdventureSupported2',

            supporterEndTime: new Date(evData.blockedTime * 1000),

            // add more pori
            supporterPories: [
              ...(mineInfo.supporterPories ?? []),
              evData.porian,
            ],
            supporterRewardLevel: [
              ...(mineInfo.supporterRewardLevel ?? []),
              evData.rewardLevel,
            ],
            supporterSlots: [...(mineInfo.supporterSlots ?? []), evData.index],
          };
        }
        break;

      case EIdleGameSCEventType.AdventureFinished:
        {
          const evData = it.data as AdventureFinishedData;
          const mineInfo = viewData.activeAdventures[evData.mineId];

          // not interesting
          if (!mineInfo) break;
          if (!(mineInfo.isSupporter || mineInfo.isFarmer)) {
            delete viewData.activeAdventures[evData.mineId];
            break;
          }

          delete viewData.activeAdventures[evData.mineId];
          const dateNo = moment(
            mineInfo.isSupporter
              ? mineInfo.supporterEndTime
              : mineInfo.farmerEndTime
          )
            .startOf('D')
            .unix();

          const previous = viewData.finishedAdventures[dateNo] || {
            unixDay: dateNo,
            timestamp: new Date(dateNo * 1000),
            finishedMineIds: [],
            totalRigy: 0,
            totalRiken: 0,
          };

          const rigyReward = mineInfo.isSupporter
            ? evData.helperReward1
            : evData.farmerReward1;
          const rikenReward = mineInfo.isSupporter
            ? evData.helperReward2
            : evData.farmerReward2;

          viewData.finishedAdventures[dateNo] = {
            ...previous,
            finishedMineIds: [...previous.finishedMineIds, mineInfo.mineId],
            totalRigy:
              previous.totalRigy +
              parseFloat(rigyReward.toString()) / TOKEN_CONVERT_CONSTANT,
            totalRiken:
              previous.totalRiken +
              parseFloat(rikenReward.toString()) / TOKEN_CONVERT_CONSTANT,
          };

          // add note
          const prevNote = viewData.note ?? {};
          viewData.note = {
            ...prevNote,
            lastMine: mineInfo.mineId,
            lastMineUnlockAt: mineInfo.isSupporter
              ? mineInfo.supporterEndTime
              : mineInfo.farmerEndTime,
          };
        }
        break;

      default:
        break;
    }
  }

  DataViewRepo.txSync(realm, () => {
    lastViewIns.cursor = allEvents[allEvents.length - 1]._id.toHexString();
    lastViewIns.data = JSON.stringify(viewData);
  });

  return viewData;
}

async function getLastView(options: IComputePlayerAdventureProps) {
  const { realm } = options;
  const viewId = computeViewId(options);

  let viewIns = await DataViewRepo.findOne(realm, viewId);
  if (!viewIns) {
    const defaultData = DataViewModel.generate(
      viewId,
      '000000000000000000000000',
      defaultViewData(options),
      HEAD_VERSION
    );
    viewIns = await DataViewRepo.createWithTx(realm, defaultData);
  }

  return viewIns;
}

function computeViewId(options: IComputePlayerAdventureProps) {
  return `player:adventure:${options.playerAddress}`;
}

function defaultViewData(
  options: IComputePlayerAdventureProps
): ActiveAdventureComputed {
  return {
    playerAddress: options.playerAddress,
    activeAdventures: {},
    finishedAdventures: {},
    note: {},
  };
}

export function humanrizeAdventureInfo(
  realm: Realm,
  advIno: AdventureInfo,
  withPoriePower = false
): AdventureInfoEx {
  const startTime = advIno.startTime
    ? new Date(advIno.startTime).toLocaleString()
    : undefined;
  const farmerEndTime = advIno.farmerEndTime
    ? new Date(advIno.farmerEndTime).toLocaleString()
    : undefined;
  const supporterEndTime = advIno.supporterEndTime
    ? new Date(advIno.supporterEndTime).toLocaleString()
    : undefined;

  let canCollect = undefined;
  const now = Date.now();
  if (advIno.isFarmer && advIno.farmerEndTime) {
    canCollect = now > new Date(advIno.farmerEndTime).valueOf();
  }
  if (advIno.isSupporter && advIno.supporterEndTime) {
    canCollect = now > new Date(advIno.supporterEndTime).valueOf();
  }

  const blockedTo = advIno.isFarmer
    ? new Date(advIno.farmerEndTime)
    : new Date(advIno.supporterEndTime as any);

  const link = `https://adventure.poriverse.io/missions/${advIno.mineId}`;
  const hasBigRewardFarmer =
    (advIno.farmerRewardLevel?.filter((itm) => itm >= 4).length ?? 0) > 0;
  const hasBigRewardSupporter =
    (advIno.supporterRewardLevel?.filter((itm) => itm >= 4).length ?? 0) > 0;

  const turnTime: any = {};
  const { farmerAtkStartAt, supporterAtkStartAt } = calculateMineTurnTime(
    new Date(advIno.startTime || '')
  );

  turnTime.farmerAtkTime = farmerAtkStartAt.toLocaleString();
  turnTime.supporterAtkTime = supporterAtkStartAt.toLocaleString();

  // calculate pories power
  const powers: Record<string, number> = {};
  if (withPoriePower) {
    const farmerPories = advIno.farmerPories ?? [];
    for (const id of farmerPories) {
      const info = PoriRepo.findOneSync(realm, id);
      if (info) powers[id] = info.minePower;
    }
    const supportPories = advIno.supporterPories ?? [];
    for (const id of supportPories) {
      const info = PoriRepo.findOneSync(realm, id);
      if (info) powers[id] = info.helpPower;
    }
  }

  return {
    link,
    canCollect,
    hasBigReward: hasBigRewardFarmer || hasBigRewardSupporter,
    ...advIno,
    startTime,
    farmerEndTime,
    supporterEndTime,
    ...turnTime,
    atkAt: advIno.isFarmer ? farmerAtkStartAt : supporterAtkStartAt,
    blockedTo,
    powers,
  };
}

export function humanrizeNote(data: ActiveAdventureComputed) {
  const note = data.note || {};
  const res: any = {
    ...note,
  };

  if (note.lastMineUnlockAt) {
    const tmp = new Date(note.lastMineUnlockAt);
    res.lastMineUnlockAt = tmp.toLocaleString();
    res.readyToStart = Date.now() > tmp.valueOf();
  }
  return res;
}

//--------------------------------------------------------------
// Note:
// all events

//  -> create
//     -> new mineObj

// -> support 1
//     -> update

//   if support != me || farmer != me => delete

// -> porify
//   if have mine -> update

// -> support 2
//   if have mine -> update

// -> finish
//   if have mine -> update + capture reward -> delete
