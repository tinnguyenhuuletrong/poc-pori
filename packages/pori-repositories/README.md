# pori-repositories

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build pori-repositories` to build the library.

## Running unit tests

Run `nx test pori-repositories` to execute the unit tests via [Jest](https://jestjs.io).
