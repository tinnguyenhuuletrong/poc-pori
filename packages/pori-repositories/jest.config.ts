module.exports = {
  displayName: 'pori-repositories',
  preset: '../../jest.preset.ts',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\.[tj]s$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/pori-repositories',
};
