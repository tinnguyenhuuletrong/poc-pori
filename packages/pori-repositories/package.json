{
  "name": "@pori-and-friends/pori-repositories",
  "version": "0.0.1",
  "type": "commonjs"
}
