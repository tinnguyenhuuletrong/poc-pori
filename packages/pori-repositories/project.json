{
  "sourceRoot": "packages/pori-repositories/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/pori-repositories",
        "main": "packages/pori-repositories/src/index.ts",
        "tsConfig": "packages/pori-repositories/tsconfig.lib.json",
        "assets": ["packages/pori-repositories/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/pori-repositories/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/pori-repositories"],
      "options": {
        "jestConfig": "packages/pori-repositories/jest.config.ts",
        "passWithNoTests": true,
        "runInBand": true
      }
    }
  },
  "tags": []
}
