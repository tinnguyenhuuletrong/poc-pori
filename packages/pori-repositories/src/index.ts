import Realm, { ConfigurationWithoutSync } from 'realm';
import { Schemas as IdleGameSchemas } from './lib/idleGames/schema';
import {
  IdleGameSCMetadataRepo,
  IdleGameSCMetadataDataModel,
  IdleGameSCEventRepo,
  IdleGameSCEventDataModel,
  PlayerRepo,
  PlayerDataModel,
  PoriRepo,
  PoriDataModel,
} from './lib/idleGames/schema';

export async function openRepo(opt: ConfigurationWithoutSync) {
  const schemas = opt.schema ?? [];
  const ins = await Realm.open({
    ...opt,
    schemaVersion: 2,
    schema: [...schemas, ...IdleGameSchemas],
  });
  return ins;
}

export {
  IdleGameSCMetadataDataModel,
  IdleGameSCEventDataModel,
  IdleGameSCMetadataRepo,
  IdleGameSCEventRepo,
  PlayerRepo,
  PlayerDataModel,
  PoriRepo,
  PoriDataModel,
};
