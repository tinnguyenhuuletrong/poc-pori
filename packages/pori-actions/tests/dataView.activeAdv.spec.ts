import * as Repos from '@pori-and-friends/pori-repositories';
import { computePlayerAdventure } from '../src/lib/dataView/playerAdventure';

describe('data_view', () => {
  let realm: Realm;

  beforeAll(async () => {
    realm = await Repos.openRepo({
      path: './_tests/testRepo.realm',
      inMemory: true,
    });
  });

  afterEach(() => {
    realm.write(() => {
      realm.deleteAll();
    });
  });

  afterAll(() => {
    realm.close();
  });

  test('active_adv_dataset_1', async () => {
    // Start -> Supported1 -> Supported2 -> Finish
    const Data = await import('./_mocks/active_adv_data_1');
    Repos.IdleGameSCEventRepo.tx(realm, () => {
      for (const it of Data.EVENTS) {
        Repos.IdleGameSCEventRepo.create(realm, it as any);
      }
    });

    const res = await computePlayerAdventure({
      playerAddress: '0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA',
      realm,
      realmEventStore: await Repos.IdleGameSCEventRepo.findAll(realm),
    });

    expect({
      res,
    }).toMatchSnapshot();
  });

  test('active_adv_dataset_2', async () => {
    // Start -> Supported1 -> Supported2
    const Data = await import('./_mocks/active_adv_data_2');
    Repos.IdleGameSCEventRepo.tx(realm, () => {
      for (const it of Data.EVENTS) {
        Repos.IdleGameSCEventRepo.create(realm, it as any);
      }
    });

    const res = await computePlayerAdventure({
      playerAddress: '0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA',
      realm,
      realmEventStore: await Repos.IdleGameSCEventRepo.findAll(realm),
    });

    expect({
      res,
    }).toMatchSnapshot();
  });

  test('active_adv_dataset_3', async () => {
    // Not related
    const Data = await import('./_mocks/active_adv_data_3');
    Repos.IdleGameSCEventRepo.tx(realm, () => {
      for (const it of Data.EVENTS) {
        Repos.IdleGameSCEventRepo.create(realm, it as any);
      }
    });

    const res = await computePlayerAdventure({
      playerAddress: '0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA',
      realm,
      realmEventStore: await Repos.IdleGameSCEventRepo.findAll(realm),
    });

    expect({
      res,
    }).toMatchSnapshot();
  });

  test('active_adv_dataset_4', async () => {
    // Not related
    const Data = await import('./_mocks/active_adv_data_4');
    Repos.IdleGameSCEventRepo.tx(realm, () => {
      for (const it of Data.EVENTS) {
        Repos.IdleGameSCEventRepo.create(realm, it as any);
      }
    });

    const res = await computePlayerAdventure({
      playerAddress: '0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA',
      realm,
      realmEventStore: await Repos.IdleGameSCEventRepo.findAll(realm),
    });

    expect({
      res,
    }).toMatchSnapshot();
  });

  test('active_adv_debug', async () => {
    // Not related
    const Data = await import('./_mocks/active_adv_data_debug');
    Repos.IdleGameSCEventRepo.tx(realm, () => {
      for (const it of Data.EVENTS) {
        Repos.IdleGameSCEventRepo.create(realm, it as any);
      }
    });

    const res = await computePlayerAdventure({
      playerAddress: '0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA',
      realm,
      realmEventStore: await Repos.IdleGameSCEventRepo.findAll(realm),
    });

    expect({
      res,
    }).toMatchSnapshot();
  });
});
