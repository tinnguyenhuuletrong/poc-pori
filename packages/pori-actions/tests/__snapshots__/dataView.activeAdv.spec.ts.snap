// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data_view active_adv_dataset_1 1`] = `
Object {
  "res": Object {
    "activeAdventures": Object {},
    "finishedAdventures": Object {
      "1651683600": Object {
        "finishedMineIds": Array [
          10530,
        ],
        "timestamp": 2022-05-04T17:00:00.000Z,
        "totalRigy": 105,
        "totalRiken": 0.567,
        "unixDay": 1651683600,
      },
    },
    "note": Object {
      "lastMine": 10530,
      "lastMineUnlockAt": 2022-05-05T04:05:20.000Z,
    },
    "playerAddress": "0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA",
  },
}
`;

exports[`data_view active_adv_dataset_2 1`] = `
Object {
  "res": Object {
    "activeAdventures": Object {
      "10530": Object {
        "farmerAddress": "0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA",
        "farmerEndTime": 2022-05-05T04:05:20.000Z,
        "farmerPories": Array [
          1876,
          1346,
          5420,
        ],
        "farmerRewardLevel": Array [
          2,
          2,
          3,
        ],
        "farmerSlots": Array [
          3,
          6,
          9,
        ],
        "isFarmer": true,
        "isSupporter": false,
        "mineId": 10530,
        "startTime": 2022-05-05T01:05:20.000Z,
        "state": "AdventureSupported2",
        "supporterAddress": "0x3048AA7A4Eb54bdc7f7b35d98603A18718380993",
        "supporterEndTime": 2022-05-05T04:05:20.000Z,
        "supporterPories": Array [
          4430,
          2768,
          2334,
          1586,
        ],
        "supporterRewardLevel": Array [
          1,
          3,
          4,
          2,
        ],
        "supporterSlots": Array [
          2,
          5,
          8,
          7,
        ],
      },
    },
    "finishedAdventures": Object {},
    "note": Object {},
    "playerAddress": "0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA",
  },
}
`;

exports[`data_view active_adv_dataset_3 1`] = `
Object {
  "res": Object {
    "activeAdventures": Object {},
    "finishedAdventures": Object {},
    "note": Object {},
    "playerAddress": "0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA",
  },
}
`;
